### PR TITLE
Update implementation of some optional arguments in Icepack

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update implementation of some optional arguments in Icepack
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Full test suite on cheyenne with 3 compilers with both Icepack, https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#67c874194a4acdaf09fc756cd679f579fa53060d, and CICE, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#b946a95ea46096d739b4b0d725d0eab81a53fbee are bit-for-bit.
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on CICE or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please provide any additional information or relevant details below:

Remove local copies where possible

Check optional arguments 

Remove public interface declarations where not needed 

Clean up some intent statements

### CODE REVIEW

**Comment by apcraig** [comment]
I think I've reviewed all the current optional arguments in Icepack now and checked implementation.  There are probably other Icepack arguments that could be made optional, but we'll need to come up with some suggestions if we want to do that.

I will merge this to the E3SM Icepack branch after it's merged and update CICE as well.

File: columnphysics/icepack_orbital.F90
Comment (phil-blain) [inline columnphysics/icepack_orbital.F90:197]: OK, so if I understand correctly these flags are not optional for CESM, so we add a check to that effect here?

If so, that makes sense.

File: columnphysics/icepack_shortwave.F90
Comment (phil-blain) [inline columnphysics/icepack_shortwave.F90:293]: OK, here we are in the call to `absorbed_solar` (not enough context is shown in this diff), and in that subroutine `fswthru_vdr` and friends are not optional, so here we have to keep the local counterparts. 

It makes sense to makes these scalars instead, nice cleanup.

File: columnphysics/icepack_shortwave.F90
Comment (phil-blain) [inline columnphysics/icepack_shortwave.F90:302]: This is a very nice cleanup because these statements are inside the condition on `aicen`. So this means that, if for some category the condition is not matched, whatever values were in `fswthru_idf` and friends stay there untouched. This was not the case before, since the whole arrays were overwritten by their local counterparts, which themselves were uninitialized. This lead to a crash in one of our configurations (when compiling with NaN initialization) which lead me to carry this commit in our internal fork: https://github.com/phil-blain/Icepack/commit/c4776379d95501d90ae79fd86d234472cd02d4f9

Nice to know I'll be able to drop this one eventually :)

File: columnphysics/icepack_therm_vertical.F90
Comment (phil-blain) [inline columnphysics/icepack_therm_vertical.F90:2459]: OK, this was already initialized below at:
https://github.com/CICE-Consortium/Icepack/blob/67c874194a4acdaf09fc756cd679f579fa53060d/columnphysics/icepack_therm_vertical.F90#L2548-L2555

but I think it makes sense to initialize it here, closer to the beginning of the subroutine. Maybe we'd want to remove that other initialization then ?

File: columnphysics/icepack_shortwave.F90
Comment (phil-blain) [inline columnphysics/icepack_shortwave.F90:72]: were the interfaces of these subroutines documented via the autodocumentation script (I did not check) ? if so, ww would want to re-run it to remove them from the public interface doc (and the same for the other subroutines that are made private in other modules).

File: columnphysics/icepack_tracers.F90
Comment (phil-blain) [inline columnphysics/icepack_tracers.F90:910]: here `icepack_write_tracer_indices` is a public subroutine so we should re-run the autodocumentation script to update its interface in the doc to reflect the change of `iounit` from optional to mandatory.

**Review by phil-blain — COMMENTED** [review]
Hi Tony,

Thanks a lot for these cleanups, these are very nice and make a lot of sense (and it's also safer, see the commit message of the commit I link to below). I left a few notes/questions below but mostly this is a direction I'm happy to see the code move towards.

Just a comment on the commit message, you mention:
> Clean up some intent statements

If I did not read too fast, that is in `columnphysics/icepack_therm_bl99.F90` and you are just moving the `intent` attributes on the same line as the argument they refer to, the actual intent (in, out or inout) does not change. I would make that clear in the commit message.

File: columnphysics/icepack_fsd.F90
Comment (apcraig) [reply columnphysics/icepack_fsd.F90:226]: You are correct, thanks for noticing this @phil-blain.  I missed that and assume it behaved like everything else in the code, didn't look closely enough.  Overall, the diagnostics should either be on all the time OR they should be turned on via the interface.  It doesn't make sense to me that they be on by default but can be turned OFF by the interface.  That's counter to how everything works in CICE/Icepack in general.  What do others think?  For now, I think I'll update the call to icepack_init_fsd_bounds and set the flag to true.

File: columnphysics/icepack_shortwave.F90
Comment (apcraig) [reply columnphysics/icepack_shortwave.F90:72]: Only the public interfaces in icepack_intfc are auto-documented.  I'll confirm that and also rerun the documentation.

File: columnphysics/icepack_therm_vertical.F90
Comment (apcraig) [reply columnphysics/icepack_therm_vertical.F90:2459]: Actually, this is odd.  In both the prior and current implementation, 

```
l_meltsliqn = c0
if (present(meltsliqn)) l_meltsliqn = meltsliqn
```
then later (line 2555 above)

`l_meltsliqn(n) = c0 
`

so I see no reason to copy meltsliqn into l_meltsliqn if it's ultimately intialized to c0.  The question is whether resetting l_meltsliqn(n) to c0 is a bug AFTER copying in meltsliqn.  Either l_meltsliqn is initialized c0 or meltsliqn.  Right now, it's confusing.  @eclare108213, any thoughts?  I may have messed this up in refactoring at some point, we should clarify and clean up.

Comment (njeffery) [reply columnphysics/icepack_therm_vertical.F90:2459]: @apcraig : Looks to me like this first passing of meltsliqn to l_meltsliqn is not needed.  The term l_meltsliqn should be zeroed out once before thermo_vertical. Line 2555 is where it's done in mpas

Comment (apcraig) [reply columnphysics/icepack_therm_vertical.F90:2459]: Thanks @njeffery, I'll clean that up.

**Comment by apcraig** [comment]
I think I've updated the changes requested.  I retested full test suite on cheyenne with Icepack and CICE, everything passes and is bit-for-bit.  This is ready for another review/merge.

File: columnphysics/icepack_orbital.F90
Comment (eclare108213) [reply columnphysics/icepack_orbital.F90:197]: This might complicate coupling in E3SM, if it's using CESMCOUPLED (I'm not sure).  @njeffery

File: columnphysics/icepack_shortwave.F90
Comment (eclare108213) [reply columnphysics/icepack_shortwave.F90:293]: Should `fswthru_vdr` and friends be optional in `absorbed_solar`?

**Review by eclare108213 — COMMENTED** [review]
Nice cleanup, thanks.  Just a couple of comments below, and I'm flagging @njeffery and @darincomeau because there are changes here to icepack_mechred, snow and orbital that would need to be considered when coupling into E3SM.

File: columnphysics/icepack_shortwave.F90
Comment (apcraig) [reply columnphysics/icepack_shortwave.F90:293]: absorbed_solar is a local routine, it's not public.  Also, it operates on 1 category while the fswthru_vdr optional arguments are arrays of size ncat.  So we cannot actually pass the full optional argument thru.  So this means we have to create a local copy for that scalar and copy in/out of the optional arguments.  We could do various things to change this like having absorbed solar operate on all categories at once.  But for now, I just left the basic implementation as it was and adapted.  Those fswthru_vdr are always passed as currently implemented so adding optional wouldn't have an impact.  What we'd need to do in that case is add a new flag to say whether to compute those or not based upon the arguments passed into shortwave_ccsm3.  That won't affect results or performance.  If absorbed_solar becomes public, then we may have to rethink this.

File: columnphysics/icepack_orbital.F90
Comment (apcraig) [reply columnphysics/icepack_orbital.F90:197]: Right, there are very few CESMCOUPLED cpps left in Icepack, we tried to clear them out a while ago and shift more to namelist settings.  Use of CESMCOUPLED in icepack_orbital was discussed of that at the time.  If this causes E3SM a problem, we can refactor it.  The problem is that it will probably require a change in all the models using CESMCOUPLED in terms of how the orbital stuff is handled, so we're trying not to change that until we have to.  But maybe we should do this at some point with consultation with CESM, RASM, etc so it's well coordinated.  New implementation still TBD as well.

**Review by eclare108213 — APPROVED** [review]
This looks good to me.  I'll approve now (because I'll be out the rest of the week) but I'd like @njeffery to take a look also.  @phil-blain does this satisfy your comments?

**Review by phil-blain — APPROVED** [review]
Looks good to me!

**Pull request merged**
